orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -963,229 +1131,461 @@ common::Status SplitGraph(Graph& graph,
+    // remove the backward sends too.
+  }
+
   graph.SetOutputs({visited_outputs.begin(), visited_outputs.end()});
   graph.SetGraphResolveNeeded();
   graph.SetGraphProtoSyncNeeded();
+  graph.Resolve();
 
-  return graph.Resolve();
+  // TODO(jufranc): once we allow partition of training graphs, we need to add 
+  // some code to make sure that the backward receive starts after the forward
+  // send, or otherwise the computation will get stuck.
+
+  return Status::OK();
 }
 
-Status ApplyPipelinePartitionToMainGraph(
-    Graph& graph,
-    const std::vector<TrainingSession::TrainingConfiguration::CutInfo>& cut_info,
-    size_t pipeline_stage_id,
-    size_t num_pipeline_stage) {
-  size_t split_count = cut_info.size();
-
-  if (num_pipeline_stage != split_count + 1) {
-    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Wrong pipeline partition cutting info. Total pipeline stage number is ",
-                           num_pipeline_stage,
-                           ", cut info length is: ",
-                           split_count);
+// Verifies the correctness of a given assignment of operators to stages.
+// Input: 
+//   - stages[i] is the stage id assigned to the operator with Index() == i.
+//   - num_stages is the total number of stages.
+//   - graph is the graph being partitioned into multiple pipeline stages.
+Status VerifyAssignment(std::vector<int> stages, int num_stages, Graph& graph) {
+
+  // All stages are used.
+  for (int s = 0; s < num_stages; ++s) {
+    auto stage_is_used = std::find(std::begin(stages), std::end(stages), s);
+    ORT_RETURN_IF_NOT(stage_is_used != std::end(stages),
+      "Stage " + std::to_string(s) + " was not assigned to any node.");
   }
 
-  std::vector<Node *> send_nodes, recv_nodes;
-  send_nodes.reserve(split_count);
-  recv_nodes.reserve(split_count);
+  // All nodes have been assigned to a stage.
+  auto op_assigned = std::find(std::begin(stages), std::end(stages), -1);
+  ORT_RETURN_IF_NOT(op_assigned == std::end(stages), 
+                    "All ops must be assigned to a stage");
 
-  // Split the graph by cutting edges specified in cut_info. After this function, the graph will be
-  // composed of several disconnected partitions.
-  ORT_RETURN_IF_ERROR(SplitGraph(graph, cut_info, send_nodes, recv_nodes));
+  // All assigned stages are within limits.
+  for (auto s : stages) {
+    ORT_RETURN_IF_NOT(s >= 0 && s < num_stages,
+                      "All stage ids must be in range.");
+  }
 
-  if (send_nodes.size() != split_count || recv_nodes.size() != split_count) {
-    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Split error: not all cut has Send and Recv inserted. Send node count: ",
-                           send_nodes.size(), ", Recv node count: ", recv_nodes.size(), ", split count: ", split_count);
+  // Edges always go forward.
+  for (size_t i = 0, t = graph.NumberOfNodes(); i < t; ++i) {
+    Node* node = graph.GetNode(i);
+    int node_stage = stages[i];
```

Comment:
>int [](start = 4, length = 3)

int -> const int. There are several places we can use `const`. Please double-check. Thanks.